### ISSUE TITLE
[C++] as-of-join backpressure for large sources

### ISSUE THREAD

**Issue by rtpsw** [apache/arrow#33640]
### Describe the enhancement requested

Add backpressure support to as-of-join in order to handle large sources. This is a merge into master of (at least most of) https://github.com/westonpace/arrow/pull/21/files

### Component(s)

C++

**Comment by rtpsw** [comment]
cc @westonpace @icexelloss

**Comment by icexelloss** [comment]
@rtpsw Do you remember what the patch is supposed to do? What about "large sources" doesn't work before the PR and how does the PR fixes that? It has been a while and I don't quite remember and the original PR doesn't seem to have too much context either.

**Comment by rtpsw** [comment]
In Acero it's the sending nodes that push data rather than the receiving nodes that pull it, so a large source can blow up the as-of-join queues (and other data structures) by pushing faster than some other source, whose lagging prevents the as-of-join node from clearing entries from memory. With this PR, as-of-join would use backpressure to hold up the faster source in order to keep memory usage within bounds.

**Comment by icexelloss** [comment]
What you are describing seems like the backpresure feature implemented in this PR:
https://github.com/westonpace/arrow/commit/45791de8311b0c2e2525e72f4c4746cc3b4364e3

What does this PR do/improve that is not in the first "AsofJoin backpresure" PR?

**Comment by westonpace** [comment]
Turns out we hadn't added https://github.com/westonpace/arrow/commit/45791de8311b0c2e2525e72f4c4746cc3b4364e3 anyways.  So I've combined both "AsofJoin backpressure" and "backpressure for large sources" into a single PR.

As for the difference:

The first PR (asofjoin backpressure) added backpressure to the asof join node.  However, when run on large sources, it tended to crash.  The main problem was the "can't shutdown plan from the process thread" problem, which has already been solved on master.  So that is why it seemed that "backpressure for large sources" wasn't adding much.

However, "backpressure for large sources" added unit tests to reproduce the issue, and these are valuable, so I've included them.
